### PR TITLE
Add new columns to the action table for #1437 and #1610

### PR DESCRIPTION
**Pull request by tlosbo**
This PR addresses issue #1437 and #1610. This adds the following columns to the `action` table:
- action_response (JSONB)
- http_status (INT)
- content_length (INT)
- sender_ip (VARCHAR(39))
- sending_org (VARCHAR(63))
- sending_org_client (VARCHAR(63))

## Changes
- Added the columns mentioned above to the action table via jooq version file.
- Updated the ActionHistory class with methods to set the columns mentioned above.
- Truncating data for `VARCHAR` columns to prevent exceptions.
- Capturing the verbose response to store in the `action_response`. We are storing the verbose response regardless of what the report POSTer specified.

## Checklist
- [ ] Submit a report then check the action table to ensure that action_response, http_status, content_length, sending_org and sending_org_client are populated as expected. The sender_ip does not get passed through locally.
- [ ] Use these headers to test the sender_ip persistence: "X-Forwarded-For: 1.2.3.4, 10.9.8.7, 1.3.5.7", "X-Azure-ClientIP: 2.4.6.8". The `X-Forwarded-For` header wins out and uses the very first one as the originating IP. If the `X-Azure-ClientIP` header is present and no `X-Forwarded-For` then that will be persisted.
- [ ] Test scenarios: pass `option=CheckConnections`, `option=ValidatePayload`, leave out the client header or use the wrong name, `verbose=true`, etc.

### Testing
- [x] Tested locally?
- [x] Ran `quickTest all`?
- [x] Ran `./prime test` against local Docker ReportStream container?
- [ ] Downloaded a file from `http://localhost:7071/api/download`?
- [ ] Added tests?

### Specific Security-related subjects a reviewer should pay specific attention to

- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?

### Process
- [ ] Are there licensing issues with any new dependencies introduced?
- [x] Includes a summary of what a code reviewer should verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

## Fixes
*List GitHub issues this PR fixes*
- #1610 
- #1437 

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #issue

### CODE REVIEW

**Comment by tlosbo** [comment]
I realized that I neglected to add/commit the versioned DDL changes to add the columns.

**Review by MauriceReeves-usds — APPROVED** [review]

File: prime-router/src/main/kotlin/azure/ActionHistory.kt [hunk 1]
```diff
@@ -163,6 +173,33 @@ class ActionHistory {
         trackActionResult(response)
     }
 
+    /**
+     * Parses the client parameter and sets the sending organization
+     * and client in the action table.
+     * @param clientParam the client header submitted with the report
+     */
+    fun trackActionSender(clientParam: String) {
+        // only set the action properties if not null
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/azure/ActionHistory.kt:182]: Should read if it is not blank.  It cannot be NULL.

File: prime-router/src/main/kotlin/azure/ActionHistory.kt [hunk 2]
```diff
@@ -163,6 +173,33 @@ class ActionHistory {
...
+        if (clientParam.isNotBlank()) {
+            try {
+                val (sendingOrg, sendingOrgClient) = Sender.parseFullName(clientParam)
+                action.sendingOrg = sendingOrg.take(ACTION.SENDING_ORG.dataType.length())
+                action.sendingOrgClient = sendingOrgClient.take(ACTION.SENDING_ORG_CLIENT.dataType.length())
+            } catch (e: Exception) {
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/azure/ActionHistory.kt:188]: There should be an action here or a comment.

File: prime-router/src/main/kotlin/azure/ReportFunction.kt [hunk 3]
```diff
@@ -135,23 +135,29 @@ class ReportFunction : Logging {
         val workflowEngine = WorkflowEngine()
         val actionHistory = ActionHistory(TaskAction.receive, context)
         var report: Report? = null
+        val verboseResponse = StringBuilder()
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/azure/ReportFunction.kt:138]: I do not see the need for this stringBuilder.  Can be just a string.

File: prime-router/src/main/kotlin/azure/ReportFunction.kt [hunk 4]
```diff
@@ -235,6 +250,13 @@ class ReportFunction : Logging {
         }
     }
 
+    /**
+     * Extract client header from request headers or query string parameters
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/azure/ReportFunction.kt:254]: Highlight request as Kdoc argument

File: prime-router/src/main/kotlin/azure/ReportFunction.kt [hunk 5]
```diff
@@ -268,7 +290,7 @@ class ReportFunction : Logging {
             .map { ResultDetail.param(ROUTE_TO_PARAMETER, "Invalid receiver name: $it") }
         errors.addAll(receiverNameErrors)
 
-        val clientName = request.headers[CLIENT_PARAMETER] ?: request.queryParameters.getOrDefault(CLIENT_PARAMETER, "")
+        val clientName = extractClientHeader(request)
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/azure/ReportFunction.kt:293]: No need to have clientName as a variable.  Use inline.

**Review by carlosfelix2 — APPROVED** [review]
Tested locally

File: prime-router/src/main/kotlin/azure/ReportFunction.kt (hunk 3 above)
Comment (MauriceReeves-usds) [reply prime-router/src/main/kotlin/azure/ReportFunction.kt:138]: I think this was future feature building. There's probably a plan to do extra building of information into the dataset.

File: prime-router/src/main/kotlin/azure/ReportFunction.kt (hunk 5 above)
Comment (MauriceReeves-usds) [reply prime-router/src/main/kotlin/azure/ReportFunction.kt:294]: I looked at this, and it does get used further down to add additional error messages.

**Review by MauriceReeves-usds — APPROVED** [review]

**Comment by MauriceReeves-usds** [comment]
Tested locally and looks good. Made a few small changes.

**Review by jimduff-usds — APPROVED** [review]
Confirmed it works locally, including the new jsonb and the IP.

Thanks, @tlosbo, your legacy lives on!!

**Pull request merged**